thorin/phase/phase.h

```diff
@@ -0,0 +1,199 @@
+#pragma once
+
+#include "thorin/def.h"
+#include "thorin/rewrite.h"
+
+#include "thorin/pass/pass.h"
+
+namespace thorin {
+
+class World;
+
+/// As opposed to a Pass, a Phase does one thing at a time and does not mix with other Phase%s.
+/// They are supposed to classically run one after another.
+/// Phase::dirty indicates whether we may need a Cleanup afterwards.
+class Phase {
+public:
+    Phase(World& world, std::string_view name, bool dirty)
+        : world_(world)
+        , name_(name)
+        , dirty_(dirty) {}
+    virtual ~Phase() = default;
+
+    /// @name getters
+    ///@{
+    World& world() { return world_; }
+    std::string_view name() const { return name_; }
+    bool is_dirty() const { return dirty_; }
+    ///@}
+
+    /// @name run/start
+    ///@{
+    virtual void run();       ///< Entry point and generates some debug output; invokes Phase::start.
+    virtual void start() = 0; ///< Actual entry.
+    ///@}
+
+    /// Runs a single Phase.
+    template<class P, class... Args>
+    static void run(Args&&... args) {
+        P p(std::forward<Args>(args)...);
+        p.run();
+    }
+
+protected:
+    World& world_;
+    std::string name_;
+    bool dirty_;
+};
+
+class RWPhase;
+
+class PhaseRewriter : public Rewriter {
+public:
+    PhaseRewriter(RWPhase& phase, World& old_world, World& new_world)
+        : Rewriter(old_world, new_world)
+        , phase(phase) {}
+
+    std::pair<const Def*, bool> pre_rewrite(const Def*) override;
+    std::pair<const Def*, bool> post_rewrite(const Def*) override;
+
+    RWPhase& phase;
+};
+
+/// Visits the current Phase::world and constructs a new World Phase::new_world along the way.
+/// It recursively **rewrites** all World::externals().
+class RWPhase : public Phase {
+public:
+    RWPhase(World& world, std::string_view name)
+        : Phase(world, name, false)
+        , new_world_(world.state())
+        , rewriter_(*this, world, new_world_) {}
+
+    void start() override;
+
+    /// @name getters
+    ///@{
+    World& old_world() { return world(); }
+    World& new_world() { return new_world_; }
+    Def2Def& old2new() { return rewriter_.old2new; }
```

Comment:
So the idea with the RWPhase is that you edit the `old2new` map upon construction of a subclass or override `{pre,post}_rewrite` to do something else than `Cleanup`?

Reply:
The idea is that you just look at the `old_world` while constructing a `new_world` along the way. Maybe, I should change that to
```c++
const World& old_world() { return world(); }
```
And yes, you would override `pre/post_rewrite`. If you just want to do a simple rewrite, you can just use the `Rewriter` for that. We can also add a `Phase` wrapper for that if this is sth that occurs regularly.